localstack/services/kinesis/kinesis_starter.py

```diff
@@ -22,28 +29,42 @@ def start_kinesis(
     :param update_listener: an update listener instance for server proxy
     :param asynchronous: currently unused but required by localstack.services.plugins.Service.start().
     TODO: either make use of this param or refactor Service.start() to not pass it.
+    :param account_id: account ID to use for this instance of Kinesis-Mock
     :returns: A running Kinesis server instance
     """
     global _server
-    if not _server:
-        _server = kinesis_mock_server.create_kinesis_mock_server(persist_path=persist_path)
 
-    _server.start()
-    log_startup_message("Kinesis")
-    port = port or config.service_port("kinesis")
+    if account_id is None:
+        account_id = get_aws_account_id()
+
+    if account_id not in _server:
+        # To support multi-accounts we use separate instance of Kinesis-Mock per account
+        # See https://github.com/etspaceman/kinesis-mock/issues/377
+        if not _server.get(account_id):
+            _server[account_id] = kinesis_mock_server.create_kinesis_mock_server(
+                account_id=account_id, persist_path=persist_path
+            )
+
+        _server[account_id].start()
+        log_startup_message("Kinesis")
+        port = port or config.service_port("kinesis")
 
-    return _server
+        check_kinesis(account_id=account_id)
```

Comment:
You basically changed the semantics of the `start_kinesis` function such that it also waits for the server to be up.
But this also means that you don't perform the check if the server has already been created (`account_id` in `_server`).
This is very dangerous, since this means that the following case won't work:
- Send a request to kinesis, which triggers the startup of the service and the backend server.
- After a few milliseconds (the server has been created but is not up yet), send another request.
  - This request will be forwarded to the backend even though it's not up yet.

So there are basically only two options:
- Avoid changing the semantics of `start_server` and just start the server asynchronously and don't wait for it to be up. Perform the check on every request (in the forwarder within the provider), as it was done before.
- Perform the check here, but implement a proper locking mechanism.

In both cases we could remove the `asynchronous` parameter though.

Reply:
Good catch, updated code to use the second option :+1: